code/game/objects/effects/glowshroom.dm

```diff
@@ -73,31 +81,49 @@
 	else //if on the floor, glowshroom on-floor sprite
 		icon_state = "[base_icon_state]f"
 
-	addtimer(CALLBACK(src, .proc/Spread), delay)
+	addtimer(CALLBACK(src, .proc/Spread), delay_spread, TIMER_UNIQUE|TIMER_NO_HASH_WAIT)
+	addtimer(CALLBACK(src, .proc/Decay), delay_decay, TIMER_UNIQUE|TIMER_NO_HASH_WAIT)	// Start decaying the plant
 
 /obj/structure/glowshroom/proc/Spread()
+	//We could be deleted at any point and the timers might not be cleaned up
+	if(QDELETED(src))
+		return
 	var/turf/ownturf = get_turf(src)
 	var/shrooms_planted = 0
-	for(var/i in 1 to myseed.yield)
-		if(prob(1/(generation * generation) * 100))//This formula gives you diminishing returns based on generation. 100% with 1st gen, decreasing to 25%, 11%, 6, 4, 2...
-			var/list/possibleLocs = list()
-			var/spreadsIntoAdjacent = FALSE
-
-			if(prob(spreadIntoAdjacentChance))
-				spreadsIntoAdjacent = TRUE
-
-			for(var/turf/simulated/floor/earth in view(3,src))
-				if(!ownturf.CanAtmosPass(earth))
-					continue
-				if(spreadsIntoAdjacent || !locate(/obj/structure/glowshroom) in view(1,earth))
-					possibleLocs += earth
-				CHECK_TICK
+	var/list/possibleLocs = list()
+	//Lets collect a list of possible viewable turfs BEFORE we iterate for yield so we don't call view multiple
+	//times when there's no real chance of the viewable range changing, really you could do this once on item
+	//spawn and most people probably would not notice.
+	for(var/turf/simulated/floor/earth in view(3,src))
+		if(is_type_in_typecache(earth, blacklisted_glowshroom_turfs))
+			continue
+		if(!ownturf.CanAtmosPass(earth))
+			continue
+		possibleLocs += earth
 
-			if(!possibleLocs.len)
+	//Lets not even try to spawn again if somehow we have ZERO possible locations
+	if(!possibleLocs.len)
+		return
+	for(var/i in 1 to myseed.yield)
+		var/chance_stats = ((myseed.potency + myseed.endurance * 2) * 0.2) // Chance of generating a new mushroom based on stats
+		var/chance_generation = (100 / (generation * generation)) // This formula gives you diminishing returns based on generation. 100% with 1st gen, decreasing to 25%, 11%, 6, 4, 2...
```

Comment:
These could probably be moved out of the loop since they don't need recalculation

Reply:
`chance_stats` does need recalculation, because in the same loop we call `proc/Decay` which lowers `endurance`. I think leaving `chance_generation` next to it for readability might be worth those few recalculations.